data/src/main/java/com/microsoft/azure/kusto/data/http/HttpClientFactory.java

```diff
@@ -35,8 +41,23 @@ public static HttpClient create(HttpClientProperties providedProperties) {
             options.setProxyOptions(properties.getProxy());
         }
 
-        // Is the per route connection maximum needed?
-        // Todo: missing keepalive and per-route connections, is keepalive handled by default in the azure clients?
+        // Is the per route connection maximum needed anymore?
+
+        // Todo: missing keepalive and per-route connections.
+        // NOTE: Keep-Alive is provided by 2 headers that can be placed on any HTTP request or response.
+        // It looks like one of the headers, Connection: Keep-Alive, is already configured when setting up tracing
```

Comment:
Thank you, the testing is definitely the part that I think would be harder for me to do on my side. I appreciate the help.

The tracing I was talking about, I think, was added by @enmoed 

There's some specific headers added for what looks like MDC (Mapped Diagnostic Context). There's a user, an app, an activity, and a request id. The request Id allows for an HTTP request to be traced from source to destination and back. Usually this is done for being able to trace a request through service logs. If you get a request with id SOME_GUID and it hits your API gateway, you can then check the logs for SOME_GUID to see that the gateway processed the request and then passed it on to its destination. Once the request reaches its destination, the request Id is then written into a response header, usually something like "x-correlation-id", indicating that you've finished the request phase, the request reached its destination and was processed, and now you can track the response on its way back to the requesting client via that correlation header. 

When adding these headers, the Connection header was also being set to "Keep-Alive", and because these HTTP tracing details were being added to every request, the Connection header was getting set on every request. However, as you mentioned, we can't really rely on Keep-Alive being set on the client because the server can just choose to ignore client Keep-Alive parameters if it's configured to do so.

For that reason, I don't know if we need to continue setting Keep-Alive details on the client-side however if we do want to continue setting those parameters, we could add the second Keep-Alive header in the request builder, so that every request gets Keep-Alive configured. We could also add a set of headers to the BaseClient which every request should integrate. Connection and Keep-Alive could be 2 of these universally applied headers.